shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Wallet.hs

```diff
@@ -109,14 +110,16 @@ getUTxO = _utxo . _utxoState . esLState . nesEs
 
 -- | Get the UTxO filtered by address.
 getFilteredUTxO ::
-  Crypto crypto =>
   ShelleyState crypto ->
   Set (Addr crypto) ->
   UTxO crypto
 getFilteredUTxO ss addrs =
-  UTxO $ Map.filter (\(TxOut addr _) -> addr `Set.member` addrs) fullUTxO
+  UTxO $ Map.filter (\(TxOutCompact addrSBS _) -> addrSBS `Set.member` addrSBSs) fullUTxO
   where
     UTxO fullUTxO = getUTxO ss
+    -- Instead of decompacting each address in the huge UTxO, compact each
+    -- address in the small set of address.
+    addrSBSs = Set.mapMonotonic (BSS.toShort . serialiseAddr) addrs
```

Comment:
Can somebody verify that `Set.mapMonotonic` is fine and that we don't need `Set.map`?

Reply:
Pretty certain it's _not_ fine. 

When we serialise addresses we include bits in the header which are determined by whether the payment/stake credentials are scripts. But these are parts of the second/third components in the address, which has a standard derived `Ord` instance.
```
putAddr (Addr network pc sr) =
  let setPayCredBit = case pc of
        ScriptHashObj _ -> flip setBit payCredIsScript
        KeyHashObj _ -> id
      netId = networkToWord8 network
   in case sr of
        StakeRefBase sc -> do
          let setStakeCredBit = case sc of
                ScriptHashObj _ -> flip setBit stakeCredIsScript
                KeyHashObj _ -> id
              header = setStakeCredBit . setPayCredBit $ netId
          B.putWord8 header
```

Reply:
Thanks for confirming this. I have updated the PR to use `Set.map`. Now it's only 8x times faster :slightly_frowning_face: